src/nvim/terminal.c

```diff
@@ -293,48 +292,49 @@ Terminal *terminal_open(TerminalOptions opts)
   return rv;
 }
 
-void terminal_close(Terminal *term, char *msg)
+/// Called when the terminal job dies
+void terminal_exit(Terminal *term, char *msg)
 {
-  if (term->closed) {
-    return;
-  }
+  assert(!term->exited);
 
   term->forward_mouse = false;
+  if (msg) {
+    terminal_receive(term, msg, strlen(msg));
+  }
 
-  // flush any pending changes to the buffer
   if (!exiting) {
+    // flush any pending changes to the buffer
     block_autocmds();
     refresh_terminal(term);
     unblock_autocmds();
   }
 
-  buf_T *buf = handle_get_buffer(term->buf_handle);
-  term->closed = true;
-
-  if (!msg || exiting) {
-    // If no msg was given, this was called by close_buffer(buffer.c).  Or if
-    // exiting, we must inform the buffer the terminal no longer exists so that
-    // close_buffer() doesn't call this again.
-    term->buf_handle = 0;
-    if (buf) {
-      buf->terminal = NULL;
-    }
-    if (!term->refcount) {
-      // We should not wait for the user to press a key.
-      term->opts.close_cb(term->opts.data);
-    }
-  } else {
-    terminal_receive(term, msg, strlen(msg));
+  terminal_destroy(term);
+}
+
+/// Called when the terminal buffer gets wiped
+void terminal_close(Terminal *term)
+{
+  if (!term->exited) {
+    term->opts.close_cb(term->opts.data);
+    terminal_destroy(term);
   }
 
-  if (buf) {
-    apply_autocmds(EVENT_TERMCLOSE, NULL, NULL, false, buf);
+  buf_T *buf = handle_get_buffer(term->buf_handle);
+  assert(buf);
+  buf->terminal = NULL;
+
+  for (size_t i = 0; i < term->sb_current; i++) {
+    xfree(term->sb_buffer[i]);
   }
+  xfree(term->sb_buffer);
+  vterm_free(term->vt);
+  xfree(term);
```

Comment:
The problem is that the vterm instance and the scrollback buffer must be kept for the lifetime of the buffer. Even after the job died, vim asks for line attributes (`terminal_get_line_attributes`) when redrawing the buffer. That function retrieves data from libvterm on the fly (which in turn uses the scrollback buffer).
What we could do is fetch the data from libvterm when the job dies and cache it to be able to free the vterm, but I don't know if that would be preferable.